File: app/src/main/java/com/sachi/mark42/databasedemo/java/UserRepo.java
package com.sachi.mark42.databasedemo.java;

import com.sachi.mark42.databaseassistant.java.DatabaseRepo;
import com.sachi.mark42.databasedemo.User;
import com.google.firebase.database.DataSnapshot;

import org.jetbrains.annotations.Nullable;

public class UserRepo extends DatabaseRepo<User> {

    private static UserRepo INSTANCE;

    @Nullable
    @Override
    public User convertDatabaseSnapshot(@Nullable DataSnapshot dataSnapshot) {
        return dataSnapshot.getValue(User.class);
    }

    public static UserRepo getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new UserRepo();
        }
        return INSTANCE;
    }
}

File: app/src/main/java/com/sachi/mark42/databasedemo/UserRepo.kt
package com.sachi.mark42.databasedemo

import com.sachi.mark42.databaseassistant.DatabaseRepo
import com.google.firebase.database.DataSnapshot

class UserRepo : DatabaseRepo<User>() {

    override fun convertDatabaseSnapshot(value: DataSnapshot?): User? {
        return value?.getValue(User::class.java)
    }

    companion object {
        private var instance : UserRepo? = null

        fun getInstance(): UserRepo {
            if (instance == null)
                instance = UserRepo()

            return instance!!
        }
        fun userPath(path: String) = "users/$path"
        fun path() = "users"
    }
}